### ISSUE TITLE
Implements idea

### ISSUE THREAD

**Issue by Thrameos** [jpype-project/jpype#330]
I had an idea to make for somewhat cleaner and safer proxy code.  The result would allow the user to implement something like this.

Assuming we have java interface defined like...
```
package com.test;
public interface MyInterface
{
  public void doSomething(int i);
}
```

Which would then be used like this in python...
```  
import jpype
from jpype import implements
import com.test.MyInterface

class MyImplementation(implements(com.test.MyInterface):
     def doSomething(self, i):
        print(i)
```

It would perform all the checking needed to make sure the interfaces were fully implemented.  And it could check that the number of arguments are met.  It would use the code for the existing proxy, so this would just be another way to do proxies that reads a bit more like java.  Thus I am not sure if it is worth it unless it is adding something.

Here is the prototype implementation.  The implementation works for both Python 2 and Python 3 series.  Anyone have an opinion on value?

```
# This produces a new proxy and verifies that the types and the methods
class _JProxyFactory(object):
    def __init__(self, interfaces):
        if not isinstance(interfaces, collections.Sequence):
            interfaces = [interfaces]
        self._interfaces = interfaces
        self._checked = False

    # Creates a new proxy for each instance
    def newProxy(self, cls, instance):
        self.checkInterfaces(cls)
        instance.__javaproxy__ =  MakeNewJProxyHere();

    # This will check to see if the interfaces are complete once per class
    def checkInterfaces(self, cls):
        if self._checked:
            return

        self._checked = True
        for interface in self._interfaces:
            if not issubclass(interface, JavaClass):
                raise TypeError("%s is not a Java interface"%interface)
            for method in interface._methods:
                if not hasattr(cls, method):
                    raise TypeError("%s does not implement method %s"%(cls.__name__,method))
                if not callable(getattr(cls, method)):
                    raise TypeError("%s does not have callable method %s"%(cls.__name___,method))

# Creates the base class instance and attaches a proxy to the new object
def _jimplements_new(self, *args):
    instance = object.__new__(self)
    instance.__javaproxy__ = self.__javaproxyfactory__.newProxy(self, instance)
    return instance

# This is a meta class generator that produces a new class for a particular set in interfaces
class _JImplements(type):
    def __new__(self, interfaces, *args):
        key = {
                '__new__':_jimplements_new,
                '__javaproxyfactory__': _JProxyFactory(interfaces) }
        # Create a new anonymous type for this set of interfaces
        return type("", (object,), key)

# implements is the public face for the type factory
implements = _JImplements
```

**Comment by Thrameos** [comment]
I am debating which is the best syntax for the implementation of this in devel.

Method 1: Dynamically creating a class base
```
class MyImplementation(implements(com.test.MyInterface):
     def doSomething(self, i):
        print(i)
```

Method 2: Using a decorator to transform a class declaration into an implementation.
```
@Implements(com.test.MyInterface)
class MyImplementation(object):
     def doSomething(self, i):
        print(i)
```

Both methods work in prototype form.  The second has the advantage that I can impose very strong checking using decorators.

```
@Implements(com.test.MyInterface)
class MyImplementation(object):
     @Override(JFloat)
     def doSomething(self, i):
        print(i)
```

gives 
```
TypeException: MyImplementation does not implement public void doSomething(int)
```
(and yes this will support multiple overloaded `@Override` )

If anyone can weigh in on what direction seems best it would be helpful.  While most of the kit I can take the general direction with object and class factories, proxies seemed very glued on.  I like the use of decorators, but I am not sure how well it is meshing with the modules design.

It may be possible with some work to pull off a dynamic byte code generator and create monstrosities like
```
@Extends(com.test.MyClass)
class MyClass(object):
     @Override(JInt)
     def doSomething(self, i):
        print("hello from python")
        # call java native MyClass implementation
        return self.super_.doSomething(i) 
```
in which we take an existing class and allow python code to be inserted arbitrarily.

**Comment by marscher** [comment]
On 07/16/2018 03:39 PM, Karl Nelson wrote:
> I am debating which is the best syntax for the implementation of this in
> devel.
> 
> Method 1: Dynamically creating a class base
> 
> |class MyImplementation(implements(com.test.MyInterface): def
> doSomething(self, i): print(i) |
> 
> Method 2: Using a decorator to transform a class declaration into an
> implementation.
> 
> |@Implements(com.test.MyInterface) class MyImplementation(object): def
> doSomething(self, i): print(i) |
> 
> Both methods work in prototype form. The second has the advantage that I
> can impose very strong checking using decorators.
> 
> |@Implements(com.test.MyInterface) class MyImplementation(object):
> @Override(JFloat) def doSomething(self, i): print(i) |
> 
> gives
> 
> |TypeException: MyImplementation does not implement public void
> doSomething(int) |
> 
> (and yes this will support multiple overloaded |@Override| )
I'm in favour of the decorator approach.
> 
> If anyone can weigh in on what direction seems best it would be helpful.
> While most of the kit I can take the general direction with object and
> class factories, proxies seemed very glued on. I like the use of
> decorators, but I am not sure how well it is meshing with the modules
> design.
> 
> It may be possible with some work to pull off a dynamic byte code
> generator and create monstrosities like
> 
> |@Extends(com.test.MyClass) class MyClass(object): @Override(JInt) def
> doSomething(self, i): print("hello from python") # call java native
> MyClass implementation return self.super_.doSomething(i) |
> 
> in which we take an existing class and allow python code to be inserted
> arbitrarily.
This would be a pretty cool feature, but the efficiency is going to be a
nightmare?

**Comment by Thrameos** [comment]
Fixed by #381
